// next-auth-sanity/credentials.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SanityCredentials = exports.signUpHandler = void 0;
const credentials_1 = __importDefault(require("next-auth/providers/credentials"));
const queries_1 = require("./queries");
const argon2_1 = __importDefault(require("argon2"));
const uuid_1 = require("@sanity/uuid");
const signUpHandler = (client, userSchema = 'user') => async (req, res) => {
    const { email, password, name, image, ...userData } = req.body;
    const user = await client.fetch(queries_1.getUserByEmailQuery, {
        userSchema,
        email
    });
    if (user) {
        res.json({ error: 'User already exist' });
        return;
    }
    const { password: _, ...newUser } = await client.create({
        _id: `user.${uuid_1.uuid()}`,
        _type: userSchema,
        email,
        password: await argon2_1.default.hash(password),
        name,
        image,
        ...userData
    });
    res.json({
        id: newUser._id,
        ...newUser
    });
};
exports.signUpHandler = signUpHandler;
const SanityCredentials = (client, userSchema = 'user') => credentials_1.default({
    name: 'Credentials',
    id: 'sanity-login',
    type: 'credentials',
    credentials: {
        email: {
            label: 'Email',
            type: 'text'
        },
        password: {
            label: 'Password',
            type: 'password'
        }
    },
    async authorize(credentials) {
        const { _id, ...user } = await client.fetch(queries_1.getUserByEmailQuery, {
            userSchema,
            email: credentials === null || credentials === void 0 ? void 0 : credentials.email
        });
        if (!user)
            throw new Error('Email does not exist');
        if (await argon2_1.default.verify(user.password, credentials === null || credentials === void 0 ? void 0 : credentials.password)) {
            return {
                id: _id,
                ...user
            };
        }
        throw new Error('Password Invalid');
    }
});
exports.SanityCredentials = SanityCredentials;
